### PR TITLE
[GlobalISel] Combine G_UNMERGE_VALUES with anyext and build vector

### PR DESCRIPTION
**Pull request by tschuett**
G_UNMERGE_VALUES (G_ANYEXT (G_BUILD_VECTOR))

ag G_UNMERGE_VALUES llvm/test/CodeGen/AArch64/GlobalISel | grep ANYEXT 
[ANYEXT] is build vector or shuffle vector

Prior art:
https://reviews.llvm.org/D87117
https://reviews.llvm.org/D87166
https://reviews.llvm.org/D87174
https://reviews.llvm.org/D87427

; CHECK-NEXT: [[BUILD_VECTOR2:%[0-9]+]]:_(<8 x s8>) = G_BUILD_VECTOR 
             [[C2]](s8), [[C2]](s8), [[C2]](s8), [[C2]](s8), [[DEF1]](s8), [[DEF1]](s8), [[DEF1]](s8), [[DEF1]](s8)

; CHECK-NEXT: [[ANYEXT1:%[0-9]+]]:_(<8 x s16>) = G_ANYEXT [[BUILD_VECTOR2]](<8 x s8>)

; CHECK-NEXT: [[UV10:%[0-9]+]]:_(<4 x s16>), [[UV11:%[0-9]+]]:_(<4 x s16>) = G_UNMERGE_VALUES 
              [[ANYEXT1]](<8 x s16>)

Test:
llvm/test/CodeGen/AArch64/GlobalISel/combine-unmerge.mir

### CODE REVIEW

File: llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h [hunk 1]
```diff
@@ -868,6 +868,14 @@ class GZext : public GCastOp {
   };
 };
 
+/// Represents an any ext.
+class GAnyExt : public GCastOp {
```
Comment (aemerson) [inline llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h:872]: You also need to add a case to the switch in `GCastOp`

File: llvm/include/llvm/CodeGen/GlobalISel/CombinerHelper.h [hunk 2]
```diff
@@ -918,6 +918,10 @@ class CombinerHelper {
   bool matchCanonicalizeICmp(const MachineInstr &MI, BuildFnTy &MatchInfo);
   bool matchCanonicalizeFCmp(const MachineInstr &MI, BuildFnTy &MatchInfo);
 
+  // unmerge_values anyext build vector
```
Comment (aemerson) [inline llvm/include/llvm/CodeGen/GlobalISel/CombinerHelper.h:921]: Please reformat/add parenthesis so it's clearer what the pattern is.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 3]
```diff
@@ -7611,3 +7611,66 @@ bool CombinerHelper::matchFoldAMinusC1PlusC2(const MachineInstr &MI,
...
+
+      MatchInfo = [=](MachineIRBuilder &B) {
+        // build into each G_UNMERGE_VALUES def
+        // a small build vector with anyext from the source build vector
+        for (unsigned I = 0; I < Unmerge->getNumDefs(); ++I) {
+          SmallVector<Register> Ops;
+          for (unsigned J = 0; J < SmallBvTy.getNumElements(); ++J) {
+            auto AnyExt = B.buildAnyExt(
+                SmallBvElemenTy,
+                BV->getSourceReg(I * SmallBvTy.getNumElements() + J));
```
Comment (aemerson) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7665]: Please re-jig this section to instead early-exit in the inverse cases when you can  to reduce indentation.

Also, comments should start with a capital letter.

**Review by aemerson — COMMENTED** [review]
Is this patch ready to be reviewed? If so, why the regressions on AMDGPU/RISCV?

File: llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h (hunk 1 above)
Comment (tschuett) [reply llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h:872]: `CastOp` already has `G_ANYZEXT` or why misunderstood your comment.

**Comment by tschuett** [comment]
```
 %and = and i48 %y, 15
  %ashr = ashr i48 %x, %and
  %trunc = trunc i48 %ashr to i16
  ret i16 %trunc
```
I bet `i48` is not a legal RISC-V type and the unmerge_values combine caused issues.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp (hunk 3 above)
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7665]: `MatchInfo` has no `return`. The `return true;` is after the `MatchInfo`. Or I misunderstood your comment.

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7665]: The indentation is owned by clang-format.

Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7665]: I would need some help here. BV->getSourceReg  is function.

Comment (arsenm) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7665]: Register Tmp = BV->getSourceReg(I * SmallBvTy.getNumElements() + J)

**Comment by tschuett** [comment]
The `trunc_shift` combine accidentally got lost. Now, the noise is limited to AArch64?!?

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 4]
```diff
@@ -7611,3 +7611,66 @@ bool CombinerHelper::matchFoldAMinusC1PlusC2(const MachineInstr &MI,
...
+
+  const MachineInstr *Source = MRI.getVRegDef(Unmerge->getSourceReg());
+
+  LLT DstTy = MRI.getType(Unmerge->getReg(0));
+
+  // We want to unmerge into vectors.
+  if (!DstTy.isFixedVector())
+    return false;
+
+  if (const GAnyExt *Any = dyn_cast<GAnyExt>(Source)) {
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7630]: Early exit and reduce indent

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp [hunk 5]
```diff
@@ -7611,3 +7611,66 @@ bool CombinerHelper::matchFoldAMinusC1PlusC2(const MachineInstr &MI,
...
+    return false;
+
+  const MachineInstr *Source = MRI.getVRegDef(Unmerge->getSourceReg());
+
+  LLT DstTy = MRI.getType(Unmerge->getReg(0));
+
+  // We want to unmerge into vectors.
+  if (!DstTy.isFixedVector())
+    return false;
+
```
Comment (arsenm) [inline llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7629]: DAG style comment for the pattern?

File: llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h (hunk 1 above)
Comment (aemerson) [reply llvm/include/llvm/CodeGen/GlobalISel/GenericMachineInstrs.h:872]: Indeed, I missed that.

**Comment by tschuett** [comment]
As shown in the summary, I found this pattern with grepping. I don't know where the LegalizerHelper creates this pattern.

File: llvm/test/CodeGen/AArch64/andorxor.ll [hunk 6]
```diff
@@ -498,21 +500,23 @@ define void @or_v4i8(ptr %p1, ptr %p2) {
...
+; CHECK-GI-NEXT:    mov v0.h[1], w8
+; CHECK-GI-NEXT:    mov v1.h[1], w9
+; CHECK-GI-NEXT:    fmov w8, s4
+; CHECK-GI-NEXT:    fmov w9, s2
+; CHECK-GI-NEXT:    mov v0.h[2], w8
+; CHECK-GI-NEXT:    mov v1.h[2], w9
+; CHECK-GI-NEXT:    fmov w8, s5
+; CHECK-GI-NEXT:    fmov w9, s3
+; CHECK-GI-NEXT:    mov v0.h[3], w8
+; CHECK-GI-NEXT:    mov v1.h[3], w9
```
Comment (aemerson) [inline llvm/test/CodeGen/AArch64/andorxor.ll:519]: Not sure what's happened here, but in general on large modern cores, copies between the GPR and FPR register banks are expensive and this code sequence will likely be slower than the original. Do you know what happened here?

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/andorxor.ll:519]: If you scroll up, the SD code looks completely different.

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/andorxor.ll:519]: Our code went from bad to even worse. Legalization and/or regbank select issues.

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/andorxor.ll:519]:
```
; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py UTC_ARGS: --version 5
; RUN: llc -O0 -mtriple=aarch64-linux-gnu -global-isel -stop-after=legalizer %s -o - | FileCheck %s

define void @or_v4i8(ptr %p1, ptr %p2) {
  ; CHECK-LABEL: name: or_v4i8
  ; CHECK: bb.1.entry:
  ; CHECK-NEXT:   liveins: $x0, $x1
  ; CHECK-NEXT: {{  $}}
  ; CHECK-NEXT:   [[COPY:%[0-9]+]]:_(p0) = COPY $x0
  ; CHECK-NEXT:   [[COPY1:%[0-9]+]]:_(p0) = COPY $x1
  ; CHECK-NEXT:   [[LOAD:%[0-9]+]]:_(s32) = G_LOAD [[COPY]](p0) :: (load (s32) from %ir.p1)
  ; CHECK-NEXT:   [[BITCAST:%[0-9]+]]:_(<4 x s8>) = G_BITCAST [[LOAD]](s32)
  ; CHECK-NEXT:   [[LOAD1:%[0-9]+]]:_(s32) = G_LOAD [[COPY1]](p0) :: (load (s32) from %ir.p2)
  ; CHECK-NEXT:   [[BITCAST1:%[0-9]+]]:_(<4 x s8>) = G_BITCAST [[LOAD1]](s32)
  ; CHECK-NEXT:   [[UV:%[0-9]+]]:_(s8), [[UV1:%[0-9]+]]:_(s8), [[UV2:%[0-9]+]]:_(s8), [[UV3:%[0-9]+]]:_(s8) = G_UNMERGE_VALUES [[BITCAST]](<4 x s8>)
  ; CHECK-NEXT:   [[DEF:%[0-9]+]]:_(s8) = G_IMPLICIT_DEF
  ; CHECK-NEXT:   [[BUILD_VECTOR:%[0-9]+]]:_(<8 x s8>) = G_BUILD_VECTOR [[UV]](s8), [[UV1]](s8), [[UV2]](s8), [[UV3]](s8), [[DEF]](s8), [[DEF]](s8), [[DEF]](s8), [[DEF]](s8)
  ; CHECK-NEXT:   [[ANYEXT:%[0-9]+]]:_(<8 x s16>) = G_ANYEXT [[BUILD_VECTOR]](<8 x s8>)
  ; CHECK-NEXT:   [[UV4:%[0-9]+]]:_(<4 x s16>), [[UV5:%[0-9]+]]:_(<4 x s16>) = G_UNMERGE_VALUES [[ANYEXT]](<8 x s16>)
  ; CHECK-NEXT:   [[UV6:%[0-9]+]]:_(s8), [[UV7:%[0-9]+]]:_(s8), [[UV8:%[0-9]+]]:_(s8), [[UV9:%[0-9]+]]:_(s8) = G_UNMERGE_VALUES [[BITCAST1]](<4 x s8>)
  ; CHECK-NEXT:   [[BUILD_VECTOR1:%[0-9]+]]:_(<8 x s8>) = G_BUILD_VECTOR [[UV6]](s8), [[UV7]](s8), [[UV8]](s8), [[UV9]](s8), [[DEF]](s8), [[DEF]](s8), [[DEF]](s8), [[DEF]](s8)
  ; CHECK-NEXT:   [[ANYEXT1:%[0-9]+]]:_(<8 x s16>) = G_ANYEXT [[BUILD_VECTOR1]](<8 x s8>)
  ; CHECK-NEXT:   [[UV10:%[0-9]+]]:_(<4 x s16>), [[UV11:%[0-9]+]]:_(<4 x s16>) = G_UNMERGE_VALUES [[ANYEXT1]](<8 x s16>)
  ; CHECK-NEXT:   [[OR:%[0-9]+]]:_(<4 x s16>) = G_OR [[UV4]], [[UV10]]
  ; CHECK-NEXT:   [[UV12:%[0-9]+]]:_(s16), [[UV13:%[0-9]+]]:_(s16), [[UV14:%[0-9]+]]:_(s16), [[UV15:%[0-9]+]]:_(s16) = G_UNMERGE_VALUES [[OR]](<4 x s16>)
  ; CHECK-NEXT:   [[DEF1:%[0-9]+]]:_(s16) = G_IMPLICIT_DEF
  ; CHECK-NEXT:   [[BUILD_VECTOR2:%[0-9]+]]:_(<8 x s16>) = G_BUILD_VECTOR [[UV12]](s16), [[UV13]](s16), [[UV14]](s16), [[UV15]](s16), [[DEF1]](s16), [[DEF1]](s16), [[DEF1]](s16), [[DEF1]](s16)
  ; CHECK-NEXT:   [[TRUNC:%[0-9]+]]:_(<8 x s8>) = G_TRUNC [[BUILD_VECTOR2]](<8 x s16>)
  ; CHECK-NEXT:   [[UV16:%[0-9]+]]:_(<4 x s8>), [[UV17:%[0-9]+]]:_(<4 x s8>) = G_UNMERGE_VALUES [[TRUNC]](<8 x s8>)
  ; CHECK-NEXT:   [[BITCAST2:%[0-9]+]]:_(s32) = G_BITCAST [[UV16]](<4 x s8>)
  ; CHECK-NEXT:   G_STORE [[BITCAST2]](s32), [[COPY]](p0) :: (store (s32) into %ir.p1)
  ; CHECK-NEXT:   RET_ReallyLR
entry:
  %d = load <4 x i8>, ptr %p1
  %e = load <4 x i8>, ptr %p2
  %s = or <4 x i8> %d, %e
  store <4 x i8> %s, ptr %p1
  ret void
}
;; NOTE: These prefixes are unused and the list is autogenerated. Do not add tests below this line:
; CHECK: {{.*}}
```

File: llvm/test/CodeGen/AArch64/GlobalISel/combine-shift-immed-mismatch-crash.mir
Comment (aemerson) [inline llvm/test/CodeGen/AArch64/GlobalISel/combine-shift-immed-mismatch-crash.mir:1]: This test update looks unrelated?

File: llvm/test/CodeGen/AArch64/andorxor.ll (hunk 6 above)
Comment (aemerson) [reply llvm/test/CodeGen/AArch64/andorxor.ll:519]: I'm not surprised that SDAG's code is completely different, and I'm not saying a regression is a deal breaker since it's overall a beneficial change. It would just be good to understand what issue there is exposed by this change on this test case, and perhaps file an issue for it so we can revisit it later.

Comment (tschuett) [reply llvm/test/CodeGen/AArch64/andorxor.ll:519]: The code after the legalizer contains the unmerge(anyext(buildvector)) pattern.

There is also unmerge(trunc(buildvector)). Maybe future work.

Note that we are before the post legalizer combiner.

File: llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp (hunk 5 above)
Comment (tschuett) [reply llvm/lib/CodeGen/GlobalISel/CombinerHelper.cpp:7646]: Is the doc sufficient?

File: llvm/test/CodeGen/AArch64/andorxor.ll (hunk 6 above)
Comment (tschuett) [reply llvm/test/CodeGen/AArch64/andorxor.ll:519]: It is also only data movement with neither computation nor loads or stores.

**Review by aemerson — APPROVED** [review]
LGTM, thanks.

**Comment by tschuett** [comment]
Thanks for review.
